python/ray/data/preprocessor.py

```diff
@@ -187,36 +188,50 @@ def _fit(self, dataset: Dataset) -> "Preprocessor":
     def _determine_transform_to_use(self, data_format: str) -> str:
         """Determine which transform to use based on data format and implementation.
 
-        * If only _transform_arrow is implemented, will convert the data to arrow.
-        * If only _transform_pandas is implemented, will convert the data to pandas.
-        If both are implemented, will pick the method corresponding to the format
-        for best performance.
-        * Implementation is defined as overriding the method in a sub-class.
+        We will infer and pick the best transform to use:
+            * ``pandas`` data format prioritizes ``pandas`` transform if available.
+            * ``arrow`` and ``numpy`` data format prioritizes ``numpy`` transform if available. # noqa: E501
+            * Fall back to what's available if no preferred path found.
         """
 
-        assert data_format in ("pandas", "arrow")
-        has_transform_arrow = (
-            self.__class__._transform_arrow != Preprocessor._transform_arrow
-        )
+        assert data_format in (
+            "pandas",
+            "arrow",
+            "numpy",
+        ), f"Unsupported data format: {data_format}"
+
         has_transform_pandas = (
             self.__class__._transform_pandas != Preprocessor._transform_pandas
         )
+        has_transform_numpy = (
+            self.__class__._transform_numpy != Preprocessor._transform_numpy
+        )
 
-        if has_transform_arrow and has_transform_pandas:
-            # Both transforms available, so we delegate based on the dataset format to
-            # ensure minimal data conversions.
-            if data_format == "pandas":
+        # Infer transform type by prioritizing native transformation to minimize
+        # data conversion cost.
+        if data_format == "pandas":
+            # Perform native pandas transformation if possible.
+            if has_transform_pandas:
                 transform_type = "pandas"
-            elif data_format == "arrow":
-                transform_type = "arrow"
-        elif has_transform_pandas:
-            transform_type = "pandas"
-        elif has_transform_arrow:
-            transform_type = "arrow"
-        else:
-            raise NotImplementedError(
-                "Neither `_transform_arrow` nor `_transform_pandas` are implemented."
-            )
+            elif has_transform_numpy:
+                transform_type = "numpy"
+            else:
+                raise NotImplementedError(
+                    "None of `_transform_numpy` or `_transform_pandas` "
+                    f"are implemented for dataset format `{data_format}`."
+                )
+        elif data_format == "arrow" or data_format == "numpy":
+            # Arrow -> Numpy is more efficient
+            if has_transform_numpy:
```

Comment:
Is this check a bit nonsensical for the `BatchMapper` since it's trivially true, i.e. it always has the `_transform_numpy` method defined even if a Pandas UDF is provided? I think that this issue is avoided if the default batch format is `"pandas"` for `BatchMapper`.